### PR TITLE
Add a .formatter.exs to generated phx.new projects

### PR DESCRIPTION
**Pull request by aptinio**
and make sure the code generated by default is formatted

### CODE REVIEW

**Comment by chrismccord** [comment]
Thank you! It looks like the following files still need some work after generating a project off this branch (Elixir 1.7.1):

```
$ mix format --check-formatted
** (Mix) mix format failed due to --check-formatted.
The following files were not formatted:

  * test/support/channel_case.ex
  * test/support/conn_case.ex
  * test/formatter_check_web/views/error_view_test.exs
  * config/dev.exs
  * config/config.exs
```

```diff
diff --git a/config/config.exs b/config/config.exs
index 5cacaf2..3b0b3e2 100644
--- a/config/config.exs
+++ b/config/config.exs
@@ -25,7 +25,6 @@ config :logger, :console,
 config :phoenix, :json_library, Jason
 config :ecto, :json_library, Jason
 
-
 # Import environment specific config. This must remain at the bottom
 # of this file so it overrides the configuration defined above.
 import_config "#{Mix.env()}.exs"
diff --git a/config/dev.exs b/config/dev.exs
index 64e26c7..5c4116a 100644
--- a/config/dev.exs
+++ b/config/dev.exs
@@ -11,8 +11,15 @@ config :formatter_check, FormatterCheckWeb.Endpoint,
   debug_errors: true,
   code_reloader: true,
   check_origin: false,
-  watchers: [node: ["node_modules/webpack/bin/webpack.js", "--mode", "development", "--watch-stdin",
-                    cd: Path.expand("../assets", __DIR__)]]
+  watchers: [
+    node: [
+      "node_modules/webpack/bin/webpack.js",
+      "--mode",
+      "development",
+      "--watch-stdin",
+      cd: Path.expand("../assets", __DIR__)
+    ]
+  ]
 
 # ## SSL Support
 #
diff --git a/test/formatter_check_web/views/error_view_test.exs b/test/formatter_check_web/views/error_view_test.exs
index d134b97..cd18655 100644
--- a/test/formatter_check_web/views/error_view_test.exs
+++ b/test/formatter_check_web/views/error_view_test.exs
@@ -9,6 +9,7 @@ defmodule FormatterCheckWeb.ErrorViewTest do
   end
 
   test "renders 500.html" do
-    assert render_to_string(FormatterCheckWeb.ErrorView, "500.html", []) == "Internal Server Error"
+    assert render_to_string(FormatterCheckWeb.ErrorView, "500.html", []) ==
+             "Internal Server Error"
   end
 end
diff --git a/test/support/channel_case.ex b/test/support/channel_case.ex
index 4b5fc13..646f99e 100644
--- a/test/support/channel_case.ex
+++ b/test/support/channel_case.ex
@@ -27,9 +27,11 @@ defmodule FormatterCheckWeb.ChannelCase do
 
   setup tags do
     :ok = Ecto.Adapters.SQL.Sandbox.checkout(FormatterCheck.Repo)
+
     unless tags[:async] do
       Ecto.Adapters.SQL.Sandbox.mode(FormatterCheck.Repo, {:shared, self()})
     end
+
     :ok
   end
 end
diff --git a/test/support/conn_case.ex b/test/support/conn_case.ex
index 96a17bb..57ec2f6 100644
--- a/test/support/conn_case.ex
+++ b/test/support/conn_case.ex
@@ -28,9 +28,11 @@ defmodule FormatterCheckWeb.ConnCase do
 
   setup tags do
     :ok = Ecto.Adapters.SQL.Sandbox.checkout(FormatterCheck.Repo)
+
     unless tags[:async] do
       Ecto.Adapters.SQL.Sandbox.mode(FormatterCheck.Repo, {:shared, self()})
     end
+
     {:ok, conn: Phoenix.ConnTest.build_conn()}
   end
 end
```
Can you take another pass? Thanks!

**Comment by aptinio** [comment]
Sure! Thanks for testing. I'll get on it.

File: installer/templates/phx_test/controllers/page_controller_test.exs [hunk 1]
```diff
@@ -2,7 +2,7 @@ defmodule <%= web_namespace %>.PageControllerTest do
   use <%= web_namespace %>.ConnCase
 
   test "GET /", %{conn: conn} do
-    conn = get conn, "/"
```
Comment (harlantwood) [inline installer/templates/phx_test/controllers/page_controller_test.exs]: I believe we want to allow `get/2` as a "DSL" call with no parens... Which I believe means it should be added to the `locals_without_parens` list in the `.formatter.exs` in the root of this repo.

I'm new here, double check. 😁

Comment (michallepicki) [reply installer/templates/phx_test/controllers/page_controller_test.exs]: @harlantwood `get/3` and `get/4` from [`Phoenix.Router`](https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/router.ex#L432)'s DSL are already added to `.formatter.exs`. This line here is in a test and uses `get/2` / `get/3` from [`Phoenix.ConnTest`](https://github.com/phoenixframework/phoenix/blob/master/lib/phoenix/test/conn_test.ex#L175), which is not a part of the Routing DSL.

**Comment by mikeover** [comment]
Thinking of working on a similar PR for umbrella projects but first, noob question: "after generating a project off this branch" - how do you generate a new Phoenix project based off a specific branch?

**Comment by michallepicki** [comment]
@mikeover You'd need to clone the repo, checkout the desired branch, and follow [installer readme](https://github.com/phoenixframework/phoenix/blob/master/installer/README.md): uninstall your current `phx_new` archive, compile and install the modified archive. Then you'd be able to use the modified `mix phx.new` installer.

**Comment by mikeover** [comment]
@michallepicki I was able to figure it out shortly after I posted that comment (typical), thank you!

**Comment by aptinio** [comment]
@chrismccord, `/test/formatter_check_web/views/error_view_test.exs` is sensitive to the length of `web_namespace` as it affects line length.

Hmmm... tricky. :thinking:

**Comment by aptinio** [comment]
@chrismccord, I generated a project using `mix phx.new phx_new` (same PATH used in test) and everything looks good.

I'm still figuring out a workaround for the line length sensitivity to base module name.

**Comment by chrismccord** [comment]
@aptinio can you rebase off of https://github.com/phoenixframework/phoenix/pull/3035 (which is now in master) so we can use this PR to handle the rest of the remaining formatter todos? Thanks!

**Comment by aptinio** [comment]
@chrismccord, so far, the following checks out:

- `mix phx.new phx_new`
- `mix phx.new phx_new --no-ecto`
- `mix phx.new phx_new --no-html`
- `mix phx.new phx_umb --umbrella`
- `mix phx.new phx_umb --umbrella --no-ecto`
- `mix phx.new phx_umb --umbrella --no-html`

I'll be tackling `phx.gen.html` and `phx.gen.json`. Should I do that here or in another PR?

**Comment by mikeover** [comment]
@aptinio Heads up, there's a new PR up #3037 that aims to handle `phx.gen.html` and `phx.gen.json`

**Comment by aptinio** [comment]
Thanks @mikeover! In that case, I'll wait and see if this PR gets merged. If that happens, maybe @tomciopp can rebase on top of this.

**Comment by aptinio** [comment]
Closing to try to get a CI rebuild..

**Comment by aptinio** [comment]
Reopening to try to get a CI rebuild...

**Comment by tomciopp** [comment]
Hmm, it looks like OTP 19.3 timed out after running `npm test`. Do you want to run again, or should we be good since all of the elixir tests have passed.

**Comment by aptinio** [comment]
We should be good now since that one passed in the first try.

**Comment by aptinio** [comment]
We should be good now because that one passed in the first try.

**Comment by aptinio** [comment]
Formatting reverted. Looks like `npm install` in the CI fails sometimes.

**Comment by aptinio** [comment]
@chrismccord, any other changes needed for this?

**Comment by chrismccord** [comment]
❤️❤️❤️🐥🔥

**Pull request merged**